File: camera/live_relay.go
package main

import (
	"io"
	"sync"
)

// LiveFrame carries a single media frame — either an H.264 NAL unit
// (video) or an Opus packet (audio). The Type field discriminates.
type LiveFrame struct {
	Data       []byte
	IsKeyframe bool // video only: true for IDR NAL units (type 5)
	IsAudio    bool // true = Opus audio, false = H.264 video
}

// LiveRelay accepts raw H.264 Annex B bytes via the io.Writer interface,
// parses NAL unit boundaries, and makes them available to consumers via
// a channel. Audio frames are pushed separately via PushAudio.
//
// The ring buffer drops oldest frames so a slow consumer never blocks
// the capture pipeline.
type LiveRelay struct {
	mu   sync.Mutex
	buf  []byte // accumulates bytes between start codes
	ring chan LiveFrame
}

// NewLiveRelay creates a relay with the given ring buffer capacity.
// A capacity of ~120 covers ~4 seconds of interleaved video+audio,
// enough for the WebSocket sender to keep up under normal conditions.
func NewLiveRelay(ringSize int) *LiveRelay {
	return &LiveRelay{
		ring: make(chan LiveFrame, ringSize),
	}
}

// C returns the channel from which consumers read parsed frames.
func (lr *LiveRelay) C() <-chan LiveFrame {
	return lr.ring
}

// PushAudio enqueues an Opus audio frame into the ring buffer.
// Called by the OGG reader goroutine, not by the io.Writer path.
func (lr *LiveRelay) PushAudio(data []byte) {
	frame := LiveFrame{
		Data:    append([]byte(nil), data...),
		IsAudio: true,
	}
	lr.enqueue(frame)
}

// Write implements io.Writer. It is called by io.MultiWriter from the
// capture pipeline tee. The bytes are raw H.264 Annex B bytestream
// (from rpicam-vid or ffmpeg).
func (lr *LiveRelay) Write(p []byte) (int, error) {
	lr.mu.Lock()
	defer lr.mu.Unlock()

	n := len(p)
	lr.buf = append(lr.buf, p...)
	lr.flush(false)
	return n, nil
}

// Close flushes any remaining buffered NAL and closes the channel.
func (lr *LiveRelay) Close() error {
	lr.mu.Lock()
	defer lr.mu.Unlock()
	lr.flush(true)
	close(lr.ring)
	return nil
}

// flush scans the buffer for Annex B start codes and emits complete
// NAL units. If final is true, the trailing data (after the last start
// code) is also emitted.
func (lr *LiveRelay) flush(final bool) {
	for {
		start := findStartCode(lr.buf, 0)
		if start < 0 {
			if final && len(lr.buf) > 0 {
				lr.emitVideo(lr.buf)
				lr.buf = nil
			}
			return
		}

		if start > 0 {
			lr.buf = lr.buf[start:]
		}

		scLen := startCodeLen(lr.buf)

		end := findStartCode(lr.buf, scLen)
		if end < 0 {
			if final {
				lr.emitVideo(lr.buf[scLen:])
				lr.buf = nil
			}
			return
		}

		lr.emitVideo(lr.buf[scLen:end])
		lr.buf = lr.buf[end:]
	}
}

// emitVideo sends a video NAL unit to the ring.
func (lr *LiveRelay) emitVideo(data []byte) {
	if len(data) == 0 {
		return
	}
	frame := LiveFrame{
		Data:       append([]byte(nil), data...),
		IsKeyframe: isIDR(data[0]),
	}
	lr.enqueue(frame)
}

// enqueue adds a frame to the ring, dropping the oldest if full.
func (lr *LiveRelay) enqueue(frame LiveFrame) {
	select {
	case lr.ring <- frame:
	default:
		// Ring full — drop oldest to make room.
		select {
		case <-lr.ring:
		default:
		}
		select {
		case lr.ring <- frame:
		default:
		}
	}
}

// findStartCode locates the next Annex B start code (0x000001 or
// 0x00000001) starting at offset. Returns -1 if not found.
func findStartCode(buf []byte, offset int) int {
	for i := offset; i < len(buf)-2; i++ {
		if buf[i] == 0 && buf[i+1] == 0 {
			if buf[i+2] == 1 {
				return i
			}
			if i+3 < len(buf) && buf[i+2] == 0 && buf[i+3] == 1 {
				return i
			}
		}
	}
	return -1
}

// startCodeLen returns 3 or 4 depending on whether the start code at
// the beginning of buf is 0x000001 or 0x00000001.
func startCodeLen(buf []byte) int {
	if len(buf) >= 4 && buf[0] == 0 && buf[1] == 0 && buf[2] == 0 && buf[3] == 1 {
		return 4
	}
	return 3
}

// isIDR returns true if the NAL unit type (encoded in the first byte
// after the start code) is an IDR slice (type 5).
func isIDR(firstByte byte) bool {
	return firstByte&0x1F == 5
}

// NullLiveRelay is an io.Writer that discards everything. Used when
// the live relay is disabled.
type NullLiveRelay struct{}

func (NullLiveRelay) Write(p []byte) (int, error) { return len(p), nil }
func (NullLiveRelay) Close() error                { return nil }
func (NullLiveRelay) PushAudio([]byte)             {}

// LiveWriter is the interface satisfied by both LiveRelay and
// NullLiveRelay, allowing the capture pipeline to tee without caring
// whether live streaming is active.
type LiveWriter interface {
	io.Writer
	Close() error
	PushAudio(data []byte)
}

File: ghostcam-py/ghostcam/live_relay.py
"""H.264 NAL Annex B parser + bounded ring buffer.

Mirrors camera/live_relay.go. Receives raw H.264 bytes via `write()`
(called from the capture pipeline tee), splits on Annex B start codes,
emits one LiveFrame per NAL unit. Audio packets are pushed separately
via `push_audio()`.

The ring is an asyncio.Queue with maxsize=ringSize. When full, the
oldest frame is dropped (matches Go's enqueue behaviour) so a slow
WebSocket consumer never back-pressures the capture pipeline.

The inner start-code scan is a single function `_find_nal_boundaries`
so a future Rust+pyo3 build can swap it via _native.find_nal_boundaries
without rewriting the asyncio plumbing. Pure Python is overwhelmingly
fast enough for the 2 Mbps Pi production rate (Spike 1 measured 883x
margin on Pi Zero 2W).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol

import asyncio

# Tunable ring depth. ~120 covers ~4 s of interleaved video+audio.
DEFAULT_RING_SIZE = 120


@dataclass(slots=True)
class LiveFrame:
    """One media frame: an H.264 NAL unit or an Opus packet."""
    data: bytes
    is_keyframe: bool = False
    is_audio: bool = False


class LiveWriter(Protocol):
    """Capture-pipeline-facing protocol satisfied by both LiveRelay and
    NullLiveRelay. Mirrors Go's LiveWriter interface."""

    def write(self, chunk: bytes) -> None: ...
    def push_audio(self, packet: bytes) -> None: ...
    def close(self) -> None: ...


class LiveRelay:
    """Annex B parser + bounded ring buffer."""

    def __init__(self, ring_size: int = DEFAULT_RING_SIZE) -> None:
        self._buf = bytearray()
        self._queue: asyncio.Queue[LiveFrame] = asyncio.Queue(maxsize=ring_size)
        self._closed = False

    @property
    def queue(self) -> asyncio.Queue[LiveFrame]:
        """Consumer reads from this queue. The WebSocket relay loops on it."""
        return self._queue

    def write(self, chunk: bytes) -> None:
        """Called from the capture-pipeline tee. Sync because the tee is
        a `loop.subprocess_exec`-driven copy task that hands off chunks
        as they come."""
        if self._closed:
            return
        self._buf.extend(chunk)
        self._flush(final=False)

    def push_audio(self, packet: bytes) -> None:
        """Enqueue an Opus audio frame. Called by the OGG reader."""
        if self._closed or not packet:
            return
        self._enqueue(LiveFrame(data=bytes(packet), is_audio=True))

    def close(self) -> None:
        if self._closed:
            return
        self._closed = True
        self._flush(final=True)
        # Sentinel: a non-blocking put attempts None — but since None
        # isn't a LiveFrame, consumers that respect the protocol stop on
        # close() being called externally. Real shutdown happens via
        # task cancellation, so we just stop accepting new frames.

    # --- internals ---

    def _flush(self, *, final: bool) -> None:
        while True:
            start = _find_start_code(self._buf, 0)
            if start < 0:
                if final and self._buf:
                    self._emit_video(bytes(self._buf))
                    self._buf = bytearray()
                return

            if start > 0:
                del self._buf[:start]

            sc_len = _start_code_len(self._buf, 0)
            end = _find_start_code(self._buf, sc_len)
            if end < 0:
                if final:
                    self._emit_video(bytes(self._buf[sc_len:]))
                    self._buf = bytearray()
                return

            self._emit_video(bytes(self._buf[sc_len:end]))
            del self._buf[:end]

    def _emit_video(self, data: bytes) -> None:
        if not data:
            return
        self._enqueue(LiveFrame(data=data, is_keyframe=_is_idr(data[0])))

    def _enqueue(self, frame: LiveFrame) -> None:
        # Drop-oldest semantics. Matches Go's two-step nonblocking
        # enqueue: try put, if full drop one and try put again.
        try:
            self._queue.put_nowait(frame)
            return
        except asyncio.QueueFull:
            pass
        try:
            self._queue.get_nowait()
        except asyncio.QueueEmpty:
            pass
        try:
            self._queue.put_nowait(frame)
        except asyncio.QueueFull:
            pass


class NullLiveRelay:
    """No-op writer. Used when live streaming is disabled."""

    def write(self, chunk: bytes) -> None: ...
    def push_audio(self, packet: bytes) -> None: ...
    def close(self) -> None: ...


# --- raw byte-level helpers (Rust-replaceable hot path) ---

_SC3 = b"\x00\x00\x01"


def _find_start_code(buf: bytes | bytearray | memoryview, offset: int) -> int:
    """Locate the next Annex B start code at offset+. Returns the index of
    the leading byte (the leading 0x00 of either 0x00000001 or 0x000001).
    -1 if none.

    Mirrors camera/live_relay.go::findStartCode. We use `bytes.find` for
    the inner scan and then promote to the 4-byte form when the byte
    immediately before the 3-byte hit (and within the offset window) is
    also 0x00. Pure Python is fast enough at Pi production rates
    (Spike 1 measured ~883 MB/s on x86, ~221 MB/s extrapolated to
    Pi Zero 2W; the camera produces ~0.25 MB/s).
    """
    pos = bytes(buf).find(_SC3, offset) if not isinstance(buf, bytes) \
        else buf.find(_SC3, offset)
    if pos < 0:
        return -1
    if pos > offset and buf[pos - 1] == 0:
        return pos - 1
    return pos


def _start_code_len(buf: bytes | bytearray | memoryview, start: int) -> int:
    """3 for 0x000001, 4 for 0x00000001. Mirrors Go's startCodeLen which
    is called on a buffer where the start code is at the beginning."""
    if start + 3 < len(buf) \
            and buf[start] == 0 and buf[start + 1] == 0 \
            and buf[start + 2] == 0 and buf[start + 3] == 1:
        return 4
    return 3


def _is_idr(first_nal_byte: int) -> bool:
    """NAL unit type is the low 5 bits of the first byte after the start
    code. Type 5 = IDR slice (keyframe)."""
    return (first_nal_byte & 0x1F) == 5
